use docker healthcheck status for readiness

tannerjfco opened the pull request:
## The Problem:
Our existing readiness check tests for a 200 status on the healthcheck endpoint of our web container. This works great for the web container, but it is not useful for determining readiness of any other container.

## The Fix:
Docker introduced `HEALTHCHECK` in 1.12. We should leverage it, so that ddev can determine the readiness of any container, so long as a `HEALTHCHECK CMD` is defined for the container.

This will help to be able to satisfy the requirement that `import-db` should not run unless the site is provisioned. Checking the db container seems like the most relevant check in this case. This will also help in writing a test case for import, as I need to wait for the db container to be ready.

## The Test:
On a test site that you have run `ddev config`, update the tags for both web and db images to the `docker-health` tag and run `ddev start`. The readiness check will use the new docker health check status. If you are using the `docker-health` tag, the site should report ready as expected.

## Automation Overview:
<!-- Please provide an overview of tests introduced by this PR, or an explanation for why no tests are needed. -->

## Related Issue Link(s):

## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->

rfay commented on Makefile:
```diff
@@ -22,7 +22,7 @@ DdevVersion ?= $(VERSION)
 WebTag ?= v0.3.4
 DBImg ?= drud/mysql-docker-local-57
-DBTag ?= v0.2.0
+DBTag ?= v0.2.1
```
DBTag [v0.3.0](https://github.com/drud/mysql-docker-local/releases/tag/v0.3.0) was already tagged by one @tannerjfco 3 days ago. Ya gotta keep up! Probably just good to rebase this PR.

rfay commented on pkg/plugins/platform/local.go:
```diff
@@ -427,3 +425,24 @@ func (l *LocalApp) AddHostsEntry() error {
 }
+
+// ContainerWait provides a wait loop to check for container in "healthy" status.
+func (l *LocalApp) ContainerWait(timeout time.Duration, containerType string) error {
```
Don't we want this in a more general location? Seems like a candidate for pkg/util/docker.go (in https://github.com/drud/ddev/pull/146

rfay commented on pkg/plugins/platform/utils.go:
```diff
@@ -394,3 +396,18 @@ func FindContainerByLabels(labels map[string]string) (docker.APIContainers, erro
+// GetContainerHealth retrieves the status of a given container. The status string returned
+// by docker contains uptime and the health status in parenths. This function will filter the uptime and
+// return only the health status.
+func GetContainerHealth(container docker.APIContainers) string {
```
I think this should go in a more general location as well.

rfay commented on pkg/version/version.go:
```diff
@@ -18,7 +18,7 @@ var WebTag = "v0.3.4" // Note that this is overridden by make
 
 // DBTag defines the default db image tag for drud dev
-var DBTag = "v0.2.0" // Note that this is overridden by make
+var DBTag = "v0.2.1" // Note that this is overridden by make
```
Whatever current is is what it should be. Not sure when 0.3.0 comes in.

rfay commented:
This still has the "Needs tests" label - is it ready to go? Since the tests explicitly use it, maybe that's it?

tannerjfco commented:
I ended up reworking and making quite a few changes after approval, so this could probably use another review. I have also addressed a couple outliers in this PR:

- Moved util sub pkgs back to util main pkg
- #123 introduced a test for ImportSQLDump, but it was disabled as it needed this work in order to succeed. It has been updated as a working test.
- I also changed app.Wait() for better flexibility to use for other containers aside from web.

beeradb commented on pkg/util/dockerutils_test.go:
```diff
@@ -0,0 +1,42 @@
+func TestGetContainerHealth(t *testing.T) {
+	assert := assert.New(t)
+	container := docker.APIContainers{
+		Status: "Up 24 seconds (health: starting)",
```
very nice.

beeradb commented on pkg/util/dockerutils.go:
```diff
@@ -175,3 +178,39 @@ func NetExists(client *docker.Client, name string) bool {
+	if strings.HasPrefix(match, pre) {
+		match = strings.TrimPrefix(match, pre)
+	}
+	return match
```
it feels a bit strange to me that there's no potential for error in any of this.
